core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1252,6 +1254,28 @@ private[spark] class DAGScheduler(
     execCores.map(cores => properties.setProperty(EXECUTOR_CORES_LOCAL_PROPERTY, cores))
   }
 
+  /**
+   * If push based shuffle is enabled, set the shuffle services to be used for the given
+   * shuffle map stage. The list of shuffle services is determined based on the list of
+   * active executors tracked by block manager master at the start of the stage.
+   */
+  private def prepareShuffleServicesForShuffleMapStage(stage: ShuffleMapStage) {
+    // TODO: Handle stage reuse/retry cases separately as without finalize changes we cannot
+    // TODO: disable shuffle merge for the retry/reuse cases
+    val mergerLocs = sc.schedulerBackend.getMergerLocations(
+      stage.shuffleDep.partitioner.numPartitions, stage.resourceProfileId)
+    logDebug(s"${stage.shuffleDep.getMergerLocs.map(_.host).mkString(", ")}")
+
+    if (mergerLocs.nonEmpty) {
+      stage.shuffleDep.setMergerLocs(mergerLocs)
+      logInfo("Shuffle merge enabled for %s (%s) with %d merger locations"
+        .format(stage, stage.name, stage.shuffleDep.getMergerLocs.size))
+    } else {
+      stage.shuffleDep.setShuffleMergeEnabled(false)
+      logInfo("Shuffle merge disabled for %s (%s)".format(stage, stage.name))
+    }
```

Comment:
QQ: Are we planning on exposing merger location is any of the spark listener events ?

Reply:
Can you please elaborate a bit on exposing merger locations through `SparkListener` events? Are you thinking about passing merger locations to `ExecutorAllocationManager` to request new executors based on this info?

Reply:
No, I meant as part of relevant `SparkListenerEvent`.
So that UI and/or analysis tools can see where the mergers are, how many merged blocks are hosted by that shuffle service, etc..

Reply:
How about doing this as part of client side metrics changes? That can introduce so many changes if we do as part of this change.

Reply:
Should we expose the list of merger locations or just a count of number of merges for a given ShuffleMapStage?
Having a long list of merge locations might increase the event size.